velox/functions/prestosql/window/NthValue.cpp

```diff
@@ -0,0 +1,189 @@
+    for (int i = 0; i < numRows; i++) {
+      if (offsetsFlatVector_->isNullAt(i)) {
+        rowNumbersVector_[i] = -1;
+      } else {
+        auto offset = offsetsFlatVector_->valueAt(i);
+        VELOX_USER_CHECK_GE(offset, 1, "Offset must be at least 1");
+        auto frameStart = frameStarts[i];
+        auto frameEnd = frameEnds[i];
+
+        auto rowNumber = frameStart + offset - 1;
+        rowNumbersVector_[i] = rowNumber <= frameEnd ? rowNumber : -1;
+      }
+    }
+  }
+
+  void apply(
+      const BufferPtr& /*peerGroupStarts*/,
+      const BufferPtr& /*peerGroupEnds*/,
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds,
+      int32_t resultOffset,
+      const VectorPtr& result) {
+    auto numRows = frameStarts->size() / sizeof(vector_size_t);
+    auto frameStartsPtr = frameStarts->as<vector_size_t>();
+    auto frameEndsPtr = frameEnds->as<vector_size_t>();
+
+    if (isConstantOffset_) {
+      getRowNumbersForOffset<true>(numRows, frameStartsPtr, frameEndsPtr);
+    } else {
+      getRowNumbersForOffset<false>(numRows, frameStartsPtr, frameEndsPtr);
+    }
+
+    auto rowNumbersRange = folly::Range(rowNumbersVector_.data(), numRows);
+    partition_->extractColumn(
+        args_[0].argIndex, rowNumbersRange, resultOffset, result);
+
+    partitionOffset_ += numRows;
+  }
+
+ private:
+  // This needs to be a copy of the function args as they could go out of
+  // scope.
+  const std::vector<exec::WindowFunctionArg> args_;
+
+  const exec::WindowPartition* partition_;
+
+  // These fields are set if the offset argument is a constant value.
+  bool isConstantOffset_ = false;
+  bool isConstantOffsetNull_ = false;
+  int64_t constantOffset_;
+
+  // This vector is used to extract values of the offset argument column
+  // (if not a constant offset value).
+  VectorPtr offsetsVector_;
+  // This is the Flat vector representation of the above vector.
+  FlatVector<int64_t>* offsetsFlatVector_;
+
+  // This offset tracks how far along the partition rows have been output.
+  // This can be used to optimize reading offset column values corresponding
+  // to the present row set in getOutput.
+  vector_size_t partitionOffset_;
+
+  // The NthValue function directly writes from the input column to the
+  // resultVector using the extractColumn API specifying the rowNumber mapping
+  // to copy between the 2 vectors. This variable is used for the rowNumber
+  // vector across getOutput calls.
+  std::vector<vector_size_t> rowNumbersVector_;
+};
+
+template <TypeKind kind>
+std::unique_ptr<exec::WindowFunction> createNthValueFunction(
+    const std::vector<exec::WindowFunctionArg>& args,
+    const TypePtr& resultType,
+    velox::memory::MemoryPool* pool) {
+  using T = typename TypeTraits<kind>::NativeType;
+  return std::make_unique<NthValueFunction<T>>(args, resultType, pool);
+}
+
+} // namespace
+
+void registerNthValue(const std::string& name) {
+  std::vector<exec::FunctionSignaturePtr> signatures{
```

Comment:
nit: Add a comment to make it easy to understand the signature

`// (T, bigint) -> T`